File: kanonak-expression/java/conformance/Conformance.java
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.kanonak.expression.Expression;
import org.kanonak.expression.Expression.ExpressionError;
import org.kanonak.expression.Expression.Ref;

/**
 * Drives the shared parity vectors through the Java kanonak-expression port - BOTH files:
 * expression-vectors.json (v1 - passes UNCHANGED under the v2 kernel; the numeric-regression
 * gate) and expression-vectors-2.json (the value-domain extension). Every vector runs through
 * evaluate AND explain and their values must agree; env bindings and expected are Values
 * (numbers, strings, arrays, {"ref": ...} objects); vectors with a trace assert the verdict
 * tree structurally.
 */
public final class Conformance {
    public static void main(String[] args) throws Exception {
        String vdir = args.length > 0 ? args[0] : "../vectors";
        int fails = run(Paths.get(vdir, "expression-vectors.json"));
        fails += run(Paths.get(vdir, "expression-vectors-2.json"));
        if (fails == 0) System.out.println("ALL VECTORS PASS");
        System.exit(fails == 0 ? 0 : 1);
    }

    /** JSON -> Value, the vector-file encoding. Booleans normalize to 1/0. */
    static Object valueOf(Object v) {
        if (v instanceof Boolean b) return b ? 1.0 : 0.0;
        if (v instanceof Number n) return n.doubleValue();
        if (v instanceof String) return v;
        if (v instanceof List<?> l) {
            List<Object> out = new ArrayList<>(l.size());
            for (Object x : l) out.add(valueOf(x));
            return out;
        }
        if (v instanceof Map<?, ?> m && m.get("ref") instanceof String r) return new Ref(r);
        throw new IllegalArgumentException("unrepresentable vector value: " + v);
    }

    /** Deep Value equality - the vector-comparison rule (lists structural). */
    static boolean deepEqual(Object a, Object b) {
        if (a instanceof List<?> al && b instanceof List<?> bl) {
            if (al.size() != bl.size()) return false;
            for (int i = 0; i < al.size(); i++) {
                if (!deepEqual(al.get(i), bl.get(i))) return false;
            }
            return true;
        }
        if (a instanceof List<?> || b instanceof List<?>) return false;
        if (a instanceof Double && b instanceof Double) return a.equals(b);
        if (a instanceof String && b instanceof String) return a.equals(b);
        if (a instanceof Ref && b instanceof Ref) return a.equals(b);
        return false;
    }

    /** The conformance context: value bindings (env) plus identity bindings (refEnv). */
    static final class Ctx {
        final Map<String, Object> env;
        final Map<String, Object> refEnv;
        Ctx(Map<String, Object> env, Map<String, Object> refEnv) { this.env = env; this.refEnv = refEnv; }
    }

    /** The conformance resolve hook: tx.VarRef -> env[varName]; any other leaf -> raise. */
    static Object resolve(Map<String, Object> node, Ctx ctx, Expression.Recurse<Ctx> evaluate) {
        if ("kanonak.org/transformations/VarRef".equals(node.get("type"))) {
            String name = (String) node.get("varName");
            if (ctx.env == null || !ctx.env.containsKey(name)) {
                throw new ExpressionError("unbound variable: " + name);
            }
            return valueOf(ctx.env.get(name));
        }
        throw new ExpressionError("unknown leaf node type: " + node.get("type"));
    }

    /** The identity-domain mirror: tx.VarRef -> refEnv[varName] member URI. */
    static String resolveRef(Map<String, Object> node, Ctx ctx) {
        if ("kanonak.org/transformations/VarRef".equals(node.get("type"))) {
            String name = (String) node.get("varName");
            if (ctx.refEnv == null || !ctx.refEnv.containsKey(name)) {
                throw new ExpressionError("unbound reference: " + name);
            }
            return (String) ctx.refEnv.get(name);
        }
        throw new ExpressionError("no reference resolver for leaf: " + node.get("type"));
    }

    /** The vector closures JSON as the typed ClosureTable shape. */
    @SuppressWarnings("unchecked")
    static Map<String, Map<String, List<String>>> closuresOf(Map<String, Object> v) {
        Object raw = v.get("closures");
        if (!(raw instanceof Map<?, ?>)) return null;
        Map<String, Map<String, List<String>>> table = new java.util.LinkedHashMap<>();
        for (Map.Entry<String, Object> prop : ((Map<String, Object>) raw).entrySet()) {
            Map<String, List<String>> inner = new java.util.LinkedHashMap<>();
            for (Map.Entry<String, Object> member : ((Map<String, Object>) prop.getValue()).entrySet()) {
                List<String> reachable = new ArrayList<>();
                for (Object m : (List<Object>) member.getValue()) reachable.add((String) m);
                inner.put(member.getKey(), reachable);
            }
            table.put(prop.getKey(), inner);
        }
        return table;
    }

    /** Structural equality of a verdict tree against the vector expected JSON tree. */
    @SuppressWarnings("unchecked")
    static boolean traceMatches(Expression.TraceNode got, Map<String, Object> want) {
        if (!got.type.equals(want.get("type"))) return false;
        if (!want.containsKey("value") || !deepEqual(got.value, valueOf(want.get("value")))) return false;
        Object wantLeft = want.get("leftRef");
        Object wantRight = want.get("rightRef");
        if (wantLeft == null ? got.leftRef != null : !wantLeft.equals(got.leftRef)) return false;
        if (wantRight == null ? got.rightRef != null : !wantRight.equals(got.rightRef)) return false;
        List<Object> wantChildren = (List<Object>) want.getOrDefault("children", List.of());
        if (wantChildren.size() != got.children.size()) return false;
        for (int i = 0; i < wantChildren.size(); i++) {
            if (!traceMatches(got.children.get(i), (Map<String, Object>) wantChildren.get(i))) return false;
        }
        return true;
    }

    @SuppressWarnings("unchecked")
    static int run(Path path) throws Exception {
        Map<String, Object> doc = (Map<String, Object>) Json.parse(Files.readString(path, StandardCharsets.UTF_8));
        List<Object> vectors = (List<Object>) doc.get("vectors");
        String name = path.getFileName().toString();
        int total = 0, pass = 0, fail = 0;
        for (Object o : vectors) {
            Map<String, Object> v = (Map<String, Object>) o;
            total++;
            String id = (String) v.get("id");
            Map<String, Object> expr = (Map<String, Object>) v.get("expr");
            Ctx ctx = new Ctx(
                (Map<String, Object>) v.getOrDefault("env", Map.of()),
                (Map<String, Object>) v.getOrDefault("refEnv", Map.of()));
            boolean expectError = Boolean.TRUE.equals(v.get("expectError"));
            Expression.EvalOptions<Ctx> options =
                new Expression.EvalOptions<>(closuresOf(v), Conformance::resolveRef);

            if (expectError) {
                boolean evalThrew = false, explainThrew = false;
                try { Expression.evaluate(expr, ctx, Conformance::resolve, options); } catch (RuntimeException e) { evalThrew = true; }
                try { Expression.explain(expr, ctx, Conformance::resolve, options); } catch (RuntimeException e) { explainThrew = true; }
                if (evalThrew && explainThrew) pass++;
                else { fail++; System.out.println("  FAIL [" + name + "/" + id + "] expected an error from evaluate AND explain"); }
                continue;
            }

            Object got;
            Expression.TraceNode trace;
            try {
                got = Expression.evaluate(expr, ctx, Conformance::resolve, options);
                trace = Expression.explain(expr, ctx, Conformance::resolve, options);
            } catch (RuntimeException e) {
                fail++; System.out.println("  FAIL [" + name + "/" + id + "] threw: " + e.getMessage());
                continue;
            }
            Object expected = valueOf(v.get("expected"));
            Object tol = v.get("tolerance");
            boolean ok;
            if (tol != null) {
                ok = got instanceof Double g && expected instanceof Double e
                    && Math.abs(g - e) <= ((Number) tol).doubleValue();
            } else {
                ok = deepEqual(got, expected);
            }
            if (!ok) {
                fail++; System.out.println("  FAIL [" + name + "/" + id + "] expected " + expected + ", got " + got);
                continue;
            }
            if (!deepEqual(trace.value, got)) {
                fail++; System.out.println("  FAIL [" + name + "/" + id + "] explain value " + trace.value + " != evaluate value " + got);
                continue;
            }
            Object wantTrace = v.get("trace");
            if (wantTrace != null && !traceMatches(trace, (Map<String, Object>) wantTrace)) {
                fail++; System.out.println("  FAIL [" + name + "/" + id + "] trace mismatch");
                continue;
            }
            pass++;
        }
        System.out.println(name + ": " + pass + "/" + total + " pass" + (fail == 0 ? "" : ", " + fail + " fail"));
        return fail;
    }

    // -- Minimal JSON parser (objects/arrays/strings/numbers/booleans/null) ----

    static final class Json {
        private final String s;
        private int i;

        private Json(String s) { this.s = s; }

        static Object parse(String s) {
            Json j = new Json(s);
            j.ws();
            Object v = j.value();
            j.ws();
            return v;
        }

        private Object value() {
            char c = s.charAt(i);
            switch (c) {
                case '{': return object();
                case '[': return array();
                case '"': return string();
                case 't': i += 4; return Boolean.TRUE;
                case 'f': i += 5; return Boolean.FALSE;
                case 'n': i += 4; return null;
                default: return number();
            }
        }

        private Map<String, Object> object() {
            Map<String, Object> m = new java.util.LinkedHashMap<>();
            i++; ws();
            if (s.charAt(i) == '}') { i++; return m; }
            while (true) {
                ws();
                String key = string();
                ws(); i++; // ':'
                ws();
                m.put(key, value());
                ws();
                char c = s.charAt(i++);
                if (c == '}') return m;
            }
        }

        private List<Object> array() {
            List<Object> a = new ArrayList<>();
            i++; ws();
            if (s.charAt(i) == ']') { i++; return a; }
            while (true) {
                ws();
                a.add(value());
                ws();
                char c = s.charAt(i++);
                if (c == ']') return a;
            }
        }

        private String string() {
            i++; // opening quote
            StringBuilder sb = new StringBuilder();
            while (true) {
                char c = s.charAt(i++);
                if (c == '"') return sb.toString();
                if (c == '\\') {
                    char e = s.charAt(i++);
                    switch (e) {
                        case '"': sb.append('"'); break;
                        case '\\': sb.append('\\'); break;
                        case '/': sb.append('/'); break;
                        case 'b': sb.append('\b'); break;
                        case 'f': sb.append('\f'); break;
                        case 'n': sb.append('\n'); break;
                        case 'r': sb.append('\r'); break;
                        case 't': sb.append('\t'); break;
                        case 'u':
                            sb.append((char) Integer.parseInt(s.substring(i, i + 4), 16));
                            i += 4;
                            break;
                        default: sb.append(e);
                    }
                } else {
                    sb.append(c);
                }
            }
        }

        private Double number() {
            int start = i;
            while (i < s.length() && "+-0123456789.eE".indexOf(s.charAt(i)) >= 0) i++;
            return Double.parseDouble(s.substring(start, i));
        }

        private void ws() {
            while (i < s.length() && Character.isWhitespace(s.charAt(i))) i++;
        }
    }
}

File: kanonak-expression/rust/tests/conformance.rs
//! Drives the shared parity vectors through the Rust kanonak-expression port —
//! BOTH files: `expression-vectors.json` (v1 — must pass UNCHANGED under the
//! v2 kernel; the numeric-regression gate) and `expression-vectors-2.json`
//! (the value-domain extension). Every vector runs through `evaluate` AND
//! `explain` and their values must agree; `env` bindings and `expected` are
//! Values (numbers, strings, arrays, `{"ref": …}` objects); vectors with a
//! `trace` assert the verdict tree structurally.

use kanonak_expression::*;
use serde_json::Value as J;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

fn vectors_dir() -> PathBuf {
    // tests run from the crate root (rust/); vectors are at ../vectors.
    let mut p = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    p.push("..");
    p.push("vectors");
    p
}

fn read(name: &str) -> J {
    let path = vectors_dir().join(name);
    serde_json::from_str(&fs::read_to_string(&path).unwrap()).unwrap()
}

/// JSON → Value, the vector-file encoding: numbers, strings, arrays (lists),
/// `{"ref": …}` objects (refs). Booleans normalize to 1/0 per the domain.
fn value_of(v: &J) -> EvalValue {
    match v {
        J::Number(n) => EvalValue::Num(n.as_f64().unwrap()),
        J::String(s) => EvalValue::Str(s.clone()),
        J::Bool(b) => EvalValue::Num(if *b { 1.0 } else { 0.0 }),
        J::Array(items) => EvalValue::List(items.iter().map(value_of).collect()),
        J::Object(o) => match o.get("ref").and_then(J::as_str) {
            Some(r) => EvalValue::Ref(r.to_string()),
            None => panic!("unrepresentable vector value: {v}"),
        },
        J::Null => panic!("null is not a Value"),
    }
}

struct Ctx {
    env: HashMap<String, EvalValue>,
    ref_env: HashMap<String, String>,
    /// A host graph for the caller's `tx.PropertyRead` leaf: ref URI → property → Value.
    graph: HashMap<String, HashMap<String, EvalValue>>,
}

/// Conformance resolve hook: a `tx.VarRef` returns `env[varName]` (error if
/// absent); a `tx.PropertyRead` is a host graph read — the documented
/// caller-leaf shape, the kernel never touches a graph. PropertyRead is the
/// reference engine's convention, pinned by the vectors so every port's harness
/// agrees: `readSource` is evaluated through the handed-back `recurse` (so a
/// loopVar bound by an enclosing iterator is visible — runtime#25) and MUST
/// yield a ref; then graph[ref][readProp] is absent → the empty list, several
/// values → a list, one value → itself. Any other unknown leaf is an error.
fn resolve_vector(
    node: &J,
    ctx: &mut Ctx,
    recurse: &mut dyn FnMut(&J, &mut Ctx) -> Result<EvalValue, ExpressionError>,
) -> Result<EvalValue, ExpressionError> {
    let typ = node.get("type").and_then(|t| t.as_str()).unwrap_or("");
    if typ == "kanonak.org/transformations/VarRef" {
        let name = node
            .get("varName")
            .and_then(|n| n.as_str())
            .ok_or_else(|| ExpressionError("VarRef missing varName".into()))?;
        match ctx.env.get(name) {
            Some(v) => Ok(v.clone()),
            None => Err(ExpressionError(format!("unbound variable '{name}'"))),
        }
    } else if typ == "kanonak.org/transformations/PropertyRead" {
        let source_node = node
            .get("readSource")
            .ok_or_else(|| ExpressionError("PropertyRead missing readSource".into()))?;
        let uri = match recurse(source_node, ctx)? {
            EvalValue::Ref(u) => u,
            other => return Err(ExpressionError(format!("PropertyRead over a non-ref: {other:?}"))),
        };
        let prop = node
            .get("readProp")
            .and_then(|p| p.as_str())
            .ok_or_else(|| ExpressionError("PropertyRead missing readProp".into()))?;
        Ok(ctx
            .graph
            .get(&uri)
            .and_then(|props| props.get(prop))
            .cloned()
            .unwrap_or(EvalValue::List(Vec::new())))
    } else {
        Err(ExpressionError(format!("unresolved leaf '{typ}'")))
    }
}

/// The identity-domain mirror: a `tx.VarRef` returns `refEnv[varName]` as a
/// member URI. Same division as `resolve_vector` — the kernel owns UriLiteral,
/// the caller owns bindings.
fn resolve_ref_vector(node: &J, ctx: &mut Ctx) -> Result<String, ExpressionError> {
    let typ = node.get("type").and_then(|t| t.as_str()).unwrap_or("");
    if typ == "kanonak.org/transformations/VarRef" {
        let name = node
            .get("varName")
            .and_then(|n| n.as_str())
            .ok_or_else(|| ExpressionError("VarRef missing varName".into()))?;
        match ctx.ref_env.get(name) {
            Some(v) => Ok(v.clone()),
            None => Err(ExpressionError(format!("unbound reference '{name}'"))),
        }
    } else {
        Err(ExpressionError(format!("no reference resolver for leaf '{typ}'")))
    }
}

fn ctx_of(v: &J) -> Ctx {
    let mut env = HashMap::new();
    if let Some(obj) = v.get("env").and_then(|e| e.as_object()) {
        for (k, val) in obj {
            env.insert(k.clone(), value_of(val));
        }
    }
    let mut ref_env = HashMap::new();
    if let Some(obj) = v.get("refEnv").and_then(|e| e.as_object()) {
        for (k, val) in obj {
            if let Some(s) = val.as_str() {
                ref_env.insert(k.clone(), s.to_string());
            }
        }
    }
    let mut graph = HashMap::new();
    if let Some(obj) = v.get("graph").and_then(|g| g.as_object()) {
        for (uri, props) in obj {
            let mut m = HashMap::new();
            if let Some(pm) = props.as_object() {
                for (prop, val) in pm {
                    m.insert(prop.clone(), value_of(val));
                }
            }
            graph.insert(uri.clone(), m);
        }
    }
    Ctx { env, ref_env, graph }
}

fn closures_of(v: &J) -> Option<ClosureTable> {
    let obj = v.get("closures")?.as_object()?;
    let mut table = ClosureTable::new();
    for (prop, members) in obj {
        let mut inner = HashMap::new();
        if let Some(m) = members.as_object() {
            for (from, reachable) in m {
                let set: Vec<String> = reachable
                    .as_array()
                    .map(|a| a.iter().filter_map(|x| x.as_str().map(String::from)).collect())
                    .unwrap_or_default();
                inner.insert(from.clone(), set);
            }
        }
        table.insert(prop.clone(), inner);
    }
    Some(table)
}

/// Deep Value equality — the vector-comparison rule (lists compare
/// structurally so a list-valued `expected` can be asserted; NOT the
/// runtime's `Equals` primitive).
fn values_deep_equal(a: &EvalValue, b: &EvalValue) -> bool {
    match (a, b) {
        (EvalValue::Num(x), EvalValue::Num(y)) => x == y,
        (EvalValue::Str(x), EvalValue::Str(y)) => x == y,
        (EvalValue::Ref(x), EvalValue::Ref(y)) => x == y,
        (EvalValue::List(x), EvalValue::List(y)) => {
            x.len() == y.len() && x.iter().zip(y.iter()).all(|(a, b)| values_deep_equal(a, b))
        }
        _ => false,
    }
}

/// Structural equality of a produced verdict tree against the vector's expected
/// JSON tree, including absent-vs-present refs.
fn trace_matches(got: &TraceNode, want: &J) -> bool {
    if want.get("type").and_then(|t| t.as_str()) != Some(got.typ.as_str()) {
        return false;
    }
    match want.get("value") {
        Some(w) => {
            if !values_deep_equal(&got.value, &value_of(w)) {
                return false;
            }
        }
        None => return false,
    }
    let want_left = want.get("leftRef").and_then(|v| v.as_str());
    if want_left != got.left_ref.as_deref() {
        return false;
    }
    let want_right = want.get("rightRef").and_then(|v| v.as_str());
    if want_right != got.right_ref.as_deref() {
        return false;
    }
    let want_children: &[J] = want
        .get("children")
        .and_then(|c| c.as_array())
        .map(|a| a.as_slice())
        .unwrap_or(&[]);
    if want_children.len() != got.children.len() {
        return false;
    }
    got.children
        .iter()
        .zip(want_children.iter())
        .all(|(g, w)| trace_matches(g, w))
}

fn run_file(name: &str) -> (usize, usize) {
    let doc = read(name);
    let vectors = doc["vectors"].as_array().unwrap();
    let total = vectors.len();
    let mut pass = 0;

    for v in vectors {
        let id = v["id"].as_str().unwrap();
        let expr = &v["expr"];
        let expect_error = v.get("expectError").and_then(|x| x.as_bool()).unwrap_or(false);
        let closures = closures_of(v);
        let options = EvalOptions {
            closures: closures.as_ref(),
            resolve_ref: Some(&resolve_ref_vector),
        };

        let mut ctx = ctx_of(v);
        let result = evaluate_with_options(expr, &mut ctx, &resolve_vector, &options);
        let mut trace_ctx = ctx_of(v);
        let traced = explain(expr, &mut trace_ctx, &resolve_vector, Some(&options));

        if expect_error {
            if result.is_err() && traced.is_err() {
                pass += 1;
            } else {
                eprintln!("FAIL [{name}/{id}] expected an error from evaluate AND explain");
            }
            continue;
        }

        let got = match result {
            Ok(got) => got,
            Err(e) => {
                eprintln!("FAIL [{name}/{id}] raised: {}", e.0);
                continue;
            }
        };
        let trace = match traced {
            Ok(t) => t,
            Err(e) => {
                eprintln!("FAIL [{name}/{id}] explain raised: {}", e.0);
                continue;
            }
        };
        let expected = value_of(&v["expected"]);
        let ok = match v.get("tolerance").and_then(|t| t.as_f64()) {
            Some(tol) => match (&got, &expected) {
                (EvalValue::Num(g), EvalValue::Num(e)) => (g - e).abs() <= tol,
                _ => false,
            },
            None => values_deep_equal(&got, &expected),
        };
        if !ok {
            eprintln!("FAIL [{name}/{id}] expected {expected:?}, got {got:?}");
            continue;
        }
        if !values_deep_equal(&trace.value, &got) {
            eprintln!(
                "FAIL [{name}/{id}] explain value {:?} != evaluate value {:?}",
                trace.value, got
            );
            continue;
        }
        if let Some(want) = v.get("trace") {
            if !trace_matches(&trace, want) {
                eprintln!("FAIL [{name}/{id}] trace mismatch");
                continue;
            }
        }
        pass += 1;
    }

    println!("{name}: {pass}/{total} pass");
    (pass, total)
}

#[test]
fn expression_vectors() {
    // v1 vectors are the regression gate: every one passes unchanged under v2.
    let (p1, t1) = run_file("expression-vectors.json");
    let (p2, t2) = run_file("expression-vectors-2.json");
    let (p3, t3) = run_align_file("expression-alignment-vectors.json");
    assert_eq!(p1, t1, "{} v1 vector(s) failed", t1 - p1);
    assert_eq!(p2, t2, "{} v2 vector(s) failed", t2 - p2);
    assert_eq!(p3, t3, "{} alignment vector(s) failed", t3 - p3);
}

/// Structural equality of an aligned tree against the vector's expected tree.
fn aligned_matches(got: &AlignedNode, want: &J) -> bool {
    let typ = want.get("type").and_then(|t| t.as_str()).unwrap_or("");
    if got.expr.get("type").and_then(|t| t.as_str()) != Some(typ) || got.trace.typ != typ {
        return false;
    }
    if got.operand.as_deref() != want.get("operand").and_then(|o| o.as_str()) {
        return false;
    }
    if got.index.map(|i| i as u64) != want.get("index").and_then(|i| i.as_u64()) {
        return false;
    }
    match (&got.element, want.get("element")) {
        (None, None) => {}
        (Some(el), Some(w)) => {
            if Some(el.loop_var.as_str()) != w.get("loopVar").and_then(|l| l.as_str()) {
                return false;
            }
            match w.get("value") {
                Some(v) if values_deep_equal(&el.value, &value_of(v)) => {}
                _ => return false,
            }
        }
        _ => return false,
    }
    match want.get("value") {
        Some(v) if values_deep_equal(&got.trace.value, &value_of(v)) => {}
        _ => return false,
    }
    let want_children: &[J] = want.get("children").and_then(|c| c.as_array()).map(|a| a.as_slice()).unwrap_or(&[]);
    if want_children.len() != got.children.len() {
        return false;
    }
    got.children.iter().zip(want_children.iter()).all(|(g, w)| aligned_matches(g, w))
}

/// The alignment vectors: explain `expr`, then `align` — the pairing must be
/// exactly the expected tree, or an error where one is expected.
fn run_align_file(name: &str) -> (usize, usize) {
    let doc = read(name);
    let vectors = doc["vectors"].as_array().unwrap();
    let mut pass = 0;
    for v in vectors {
        let id = v["id"].as_str().unwrap();
        let expr = &v["expr"];
        let closures = closures_of(v);
        let options = EvalOptions { closures: closures.as_ref(), resolve_ref: Some(&resolve_ref_vector) };
        let mut ctx = ctx_of(v);
        let trace = match explain(expr, &mut ctx, &resolve_vector, Some(&options)) {
            Ok(t) => t,
            Err(e) => {
                println!("{name}/{id}: explain failed: {}", e.0);
                continue;
            }
        };
        if v.get("expectError").and_then(|x| x.as_bool()).unwrap_or(false) {
            let target = v.get("alignExpr").unwrap_or(expr);
            if align(target, &trace).is_ok() {
                println!("{name}/{id}: expected align to reject the trace");
            } else {
                pass += 1;
            }
            continue;
        }
        match align(expr, &trace) {
            Ok(got) => {
                if aligned_matches(&got, &v["expected"]) {
                    pass += 1;
                } else {
                    println!("{name}/{id}: alignment mismatch");
                }
            }
            Err(e) => println!("{name}/{id}: align failed: {}", e.0),
        }
    }
    println!("{name}: {pass}/{} pass", vectors.len());
    (pass, vectors.len())
}
